### ISSUE TITLE
Make compatible with webpack-dev-server 4.0 beta.

### ISSUE THREAD

**Issue by anark** [rails/webpacker#2994]
webpack-dev-server has a new beta release(4.0.0-beta.2) that features built in webpack 5 support.

Would be great to make webpacker 6.0 compatible with this as well
https://github.com/webpack/webpack-dev-server/blob/master/CHANGELOG.md#400-beta0-2020-11-27

**Comment by vladimirtemnikov** [comment]
And it has much less dependencies ;)

**Comment by navidemad** [comment]
Is there any news about this issue :) ?

**Comment by jcoyne** [comment]
Yes, #3057 is one thing that has to be done. Another issue I ran into is that `watchOptions` (https://github.com/rails/webpacker/blob/032c2d11a3758035305fe34e6795ab133128ee32/package/environments/development.js#L48) isn't supported in webpack-dev-server 4.

**Comment by clinejj** [comment]
Here's my `config/webpack/development.js`  that I got working with webpack-dev-server 4.0beta3 that maps the current configuration onto the new configuration:

```js
process.env.NODE_ENV = process.env.NODE_ENV || 'development'

const webpackConfig = require('./base')

const devServerConfig = {
  client: {
    logging: webpackConfig.devServer.clientLogLevel,
    overlay: webpackConfig.devServer.overlay,
    needClientEntry: webpackConfig.devServer.injectClient,
    needHotEntry: webpackConfig.devServer.injectHot,
  },
  compress: webpackConfig.devServer.compress,
  devMiddleware: {
    publicPath: webpackConfig.devServer.publicPath,
  },
  firewall: !webpackConfig.devServer.disableHostCheck,
  host: webpackConfig.devServer.host,
  port: webpackConfig.devServer.port,
  https: webpackConfig.devServer.https,
  hot: webpackConfig.devServer.hot,
  historyApiFallback: webpackConfig.devServer.historyApiFallback,
  headers: webpackConfig.devServer.headers,
  static: [
    {
      directory: webpackConfig.devServer.contentBase,
      watch: true,
    }
  ],
};
if (webpackConfig.devServer.quiet) {
  webpackConfig.infrastructureLogging.level = 'none';
}
if (webpackConfig.devServer.useLocalIp) {
  devServerConfig.host = 'local-ip';
}
webpackConfig.stats = webpackConfig.devServer.stats;
webpackConfig.devServer = devServerConfig;

module.exports = webpackConfig;
```

It's a bit tricky to parse the changed configurations from the [changelog](https://github.com/webpack/webpack-dev-server/blob/f2f335c16a423d61edc974d46377afc1a649d88a/CHANGELOG.md) for wds, but this should handle it. If you have custom pack paths or a more advanced configuration you may want to verify that everything is being set correctly.

**Comment by tagliala** [comment]
webpack-dev-server [entered the RC version stage](https://github.com/webpack/webpack-dev-server/blob/master/CHANGELOG.md#400-rc0-2021-07-19)

**Comment by anark** [comment]
Webpack Dev Server 4 has now been released as a stable 4.0.0 release.   What needs to be done to make webpacker compatible with webpack-dev-server@4.0.0

**Comment by clinejj** [comment]
The patch I posted above still works if you want to upgrade sooner, but when I briefly looked into this earlier rails/webpacker would need to update the following:

- update this file to match the new configuration and options available in 4: https://github.com/rails/webpacker/blob/032c2d11a3758035305fe34e6795ab133128ee32/package/environments/development.js
- I think some of the methods in this class would need to be adjusted based on some of the changed/removed configurations: https://github.com/rails/webpacker/blob/032c2d11a3758035305fe34e6795ab133128ee32/lib/webpacker/dev_server.rb#L1
- the YML configuration would need to be updated with the new options as well: https://github.com/rails/webpacker/blob/032c2d11a3758035305fe34e6795ab133128ee32/lib/install/config/webpacker.yml#L23

I don't have enough of an understanding on how the YML configuration gets passed through to the JS but I think those are the three main files that handle it.

**Comment by PikachuEXE** [comment]
There is a [JS config file](https://github.com/rails/webpacker/blob/v6.0.0.rc.1/package/config.js#L9) that read the "default config" from [a YAML file with default config](https://github.com/rails/webpacker/blob/v6.0.0.rc.1/lib/install/config/webpacker.yml)
Then the same config file [read from your project's config file and merge values](https://github.com/rails/webpacker/blob/v6.0.0.rc.1/package/config.js#L11-L19)

So to get rid of the unwanted options either the user need an extra statement to remove it or update the default config file from `webpack-dev-server`

**Comment by RSO** [comment]
> Here's my `config/webpack/development.js` that I got working with webpack-dev-server 4.0rc0 that maps the current configuration onto the new configuration:
> 
> NOTE: updated Aug 2 to support weback-dev-server 4.0rc0
> 
> ```js
> process.env.NODE_ENV = process.env.NODE_ENV || 'development'
> 
> const webpackConfig = require('./base')
> 
> const devServerConfig = {
>   client: {
>     logging: webpackConfig.devServer.clientLogLevel,
>     overlay: webpackConfig.devServer.overlay,
>   },
>   compress: webpackConfig.devServer.compress,
>   devMiddleware: {
>     publicPath: webpackConfig.devServer.publicPath,
>   },
>   hot: webpackConfig.devServer.injectHot,
>   host: webpackConfig.devServer.host,
>   port: webpackConfig.devServer.port,
>   https: webpackConfig.devServer.https,
>   hot: webpackConfig.devServer.hot,
>   historyApiFallback: webpackConfig.devServer.historyApiFallback,
>   headers: webpackConfig.devServer.headers,
>   static: [
>     {
>       directory: webpackConfig.devServer.contentBase,
>       watch: true,
>     }
>   ],
> };
> if (webpackConfig.devServer.quiet) {
>   webpackConfig.infrastructureLogging.level = 'none';
> }
> if (webpackConfig.devServer.useLocalIp) {
>   devServerConfig.host = 'local-ip';
> }
> if (webpackConfig.devServer.disableHostCheck) {
>   devServerConfig.allowedHosts = 'all';
> }
> webpackConfig.stats = webpackConfig.devServer.stats;
> webpackConfig.devServer = devServerConfig;
> 
> module.exports = webpackConfig;
> ```
> 
> It's a bit tricky to parse the changed configurations from the [changelog](https://github.com/webpack/webpack-dev-server/blob/master/CHANGELOG.md) for wds, but this should handle it. If you have custom pack paths or a more advanced configuration you may want to verify that everything is being set correctly.

Thanks for this! Just a small note for others trying this same snippet:
1. `hot:` is set twice in the `devServerConfig` variable
2. I changed `watch: true` to `watch: webpackConfig.devServer.watchOptions` so that it (hopefully) includes my `ignores: /node_modules/` in there!

**Comment by anark** [comment]
Thanks @RSO,
Those changes do work for me when I try to run it(as long as I remove react-refresh but I'll tackle that later).

However It would be better if webpacker users didn't have to use a custom config to use the latest webpack-dev-server.  I think https://github.com/rails/webpacker/pull/3122 seems like a reasonable solution.  That way when upgrading to webpacker 6.0 you would also upgrade to dev-server 4.0

**Comment by PikachuEXE** [comment]
Agree with @anark 
`webpacker` 6.0 stable release is not made yet and it's fine we include more breaking changes.
Just need the migration doc to catch up

**Comment by PikachuEXE** [comment]
For those who want to try out #3122 I posted [a comment in that PR](https://github.com/rails/webpacker/pull/3122#issuecomment-902407580)
about the method, the result and a potential issue for `webpacker`/`webpack-dev-server` v4

**Comment by dhh** [comment]
Solved via #3122.
